### PR TITLE
log-vm: To centralize logs using grafana promtail & loki

### PR DESCRIPTION
**Pull request by vunnyso**
## Checklist for things done

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [ ] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `nix flake check --accept-flake-config` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [ ] Author has added reviewers and removed PR draft status

## Testing

**To view logs in locally on ghaf target machine**

ssh into log-vm and use `logcli` to view systemd journal logs
```
[ghaf@ghaf-host:~]$ ssh 192.168.101.66

[ghaf@log-vm:~]$ logcli query '{job="systemd-journal"}'

[ghaf@log-vm:~]$ logcli query '{hostname="ghaf-host"}'
```

**To view logs in grafana GUI on remote server/ dev machine.**

1) Move to dev machine / machine where you want to install `grafana`
 To install grafana on NixOS you can follow mentioned guide [here](https://nixos.wiki/wiki/Grafana) 
 
 **Example:**
```networking.firewall.allowedTCPPorts = [ 3000 ];
services.grafana = {
  enable   = true;
  settings = {
    server = {
      http_addr = "<your_local_network_grafana_instance_ip_address_here>";
      domain = "localdomain";
      http_port = 3000;
      protocol = "http";
    };
  };

  dataDir  = "/var/lib/grafana";
};
```
2) Make sure you have connected to WIFI using `nm-launcher` on `ghaf target` machine.

3) Run the below command to add `loki` data source, replace with `netvm-wifi-ip-address` with yours IP.
  [ghaf@log-vm:~]$ curl 'http://admin:admin@<grafana_ip_in_your_local_network>:3000/api/datasources'
-X POST -H 'Content-Type: application/json;charset=UTF-8' --data-binary '{"name":"loki","type":"loki","url":"http://<net-vm_wifi_ip_address>:3100","access":"proxy","isDefault":true,"database":"tsdb"}'

4) Go to browser where you want to view the logs and type following address `http://<grafana_ip_in_your_local_network>:3000/explore`
  Follow below steps in order to view logs/download logs.

image

Also there is way to filter out logs based on hostName as in example below:
image

## Next Steps
1) Build Grafana alloy nix package
2) Change from promtail agent to alloy

### CODE REVIEW

**Comment by vunnyso** [comment]
> Better link testing instructions for grafana installation to https://nixos.wiki/wiki/Grafana instead of https://community.grafana.com/t/installing-on-nixos/6712 The latter link is from 2018 and many nixos grafana settings have been changed - for example
> 
> ```
> trace: warning: The option `services.grafana.domain' defined in `/nix/store/mzgmjayar7pm38x46yxhf27q4jv11zmp-source/hosts/nixos/carrie/configuration.nix' has been renamed to `services.grafana.settings.server.domain'.
> trace: warning: The option `services.grafana.port' defined in `/nix/store/mzgmjayar7pm38x46yxhf27q4jv11zmp-source/hosts/nixos/carrie/configuration.nix' has been renamed to `services.grafana.settings.server.http_port'.
> trace: warning: The option `services.grafana.protocol' defined in `/nix/store/mzgmjayar7pm38x46yxhf27q4jv11zmp-source/hosts/nixos/carrie/configuration.nix' has been renamed to `services.grafana.settings.server.protocol'.
> ```
> 
> I got grafana running with:
> 
> ```
> networking.firewall.allowedTCPPorts = [ 3000 ];
> services.grafana = {
>   enable   = true;
>   settings = {
>     server = {
>       http_addr = "<your_local_network_grafana_instance_ip_address_here>";
>       domain = "localdomain";
>       http_port = 3000;
>       protocol = "http";
>     };
>   };
> 
>   dataDir  = "/var/lib/grafana";
> };
> ```

Thanks @vilvo  for details, I have updated `Testing` section accordingly.

**Comment by vilvo** [comment]
Also, please change step 3. in Testing section instructions to:
```
[ghaf@log-vm:~]$ curl 'http://admin:admin@<grafana_ip_in_your_local_network>:3000/api/datasources'
-X POST -H 'Content-Type: application/json;charset=UTF-8' --data-binary '{"name":"loki","type":"loki","url":"http://<net-vm_wifi_ip_address>:3100","access":"proxy","isDefault":true,"database":"tsdb"}'
```
As of now, it's unclear where the tester is expected to run the command. Also, it's better use grafana instance ip address instead of localhost. Tested it to work that way. Also, the ghaf logs filtering based on the ghaf hostname works fine 👍🏻

File: modules/microvm/virtualization/microvm/logvm.nix [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+              }
+            ];
+            linkConfig.ActivationPolicy = "always-up";
+          };
+        };
+
+        networking.firewall.allowedTCPPorts = [3100];
+
+        environment.etc."loki.yaml".source = ./loki-local-config.yaml;
+        systemd.services.loki = {
```
Comment (Mic92) [inline modules/microvm/virtualization/microvm/logvm.nix:60]: Btw. I had to some fixes in my configuration after the last loki upgrade in nixpkgs unstable: https://github.com/Mic92/dotfiles/blob/main/nixos/eva/modules/loki.nix

File: modules/microvm/virtualization/microvm/loki-local-config.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+# SPDX-License-Identifier: Apache-2.0
+auth_enabled: false
+
+server:
+  http_listen_port: 3100
+  grpc_listen_port: 9096
+
+common:
+  instance_addr: 127.0.0.1
+  path_prefix: /tmp/loki
```
Comment (Mic92) [inline modules/microvm/virtualization/microvm/loki-local-config.yaml:11]: I would probably still put it in `/var/lib/loki` also this VM doesn't realistically has any other services /tmp is still world-readable after all. In my experience people might end up copying this at some point to a place where this matters.

File: modules/microvm/virtualization/microvm/logvm.nix (hunk 1 above)
Comment (vunnyso) [reply modules/microvm/virtualization/microvm/logvm.nix:60]: Thanks @Mic92 for pointing to correct settings.

**Comment by vilvo** [comment]
Thanks Vunny for working with this. Can you please do major refactorings to this PR according to what has already been agreed, communicated and implemented in https://github.com/tiiuae/ghaf-givc ?

To summarise:
- refactor promtail & loki into a nix module instead of directly declared in log-vm microvm.nix declaration. Please refer to ghaf-givc admin-module for example and reference. 
- remove all log-vm - the centralised logging functionality module will be imported in reference admin-vm (already merged)
- open a new draft PR to ghaf-givc with updated PR title and testing instructions for the centralised logging module parts
- change this PR to integrate the ghaf-givc logging module from previous steps to ghaf admin-vm

Please let me know if you have questions.

File: modules/microvm/virtualization/microvm/loki-local-config.yaml (hunk 2 above)
Comment (vilvo) [reply modules/microvm/virtualization/microvm/loki-local-config.yaml:11]: Sorry, prev comment still applies structurally but for `alloy`, not for `promtail` - I should be more focused but I wasn't. My excuse is Friday and the heat making my brain soft ...

**Comment by vunnyso** [comment]
Closing this draft PR since we are not planning to integrate promtail module, instead will use grafana alloy in https://github.com/tiiuae/ghaf/pull/625.
Will re-open if required in future.

File: modules/microvm/virtualization/microvm/loki-local-config.yaml (hunk 2 above)
Comment (vunnyso) [reply modules/microvm/virtualization/microvm/loki-local-config.yaml:11]: Hi @vilvo  I have addressed your review comments in https://github.com/tiiuae/ghaf/pull/625
Please review it.

**Pull request closed without merging**